components/renku_data_services/utils/core.py

```diff
@@ -90,3 +91,71 @@ async def transaction_wrapper(self: _WithSessionMaker, *args: _P.args, **kwargs:
             return await f(self, *args, **kwargs)
 
     return transaction_wrapper
+
+
+def _get_url(host: str) -> str:
+    return f"https://{host}/openbis/openbis/rmi-application-server-v3.json"
+
+
+async def get_openbis_session_token(
+    host: str,
+    username: str,
+    password: str,
+    timeout: int = 12,
+) -> str:
+    """Requests an openBIS session token with the user's login credentials."""
+    login = {"method": "login", "params": [username, password], "id": "2", "jsonrpc": "2.0"}
+    async with httpx.AsyncClient(verify=get_ssl_context()) as client:
+        response = await client.post(_get_url(host), json=login, timeout=timeout)
+        json: dict[str, str] = response.json()
+        return json["result"]
+
+
+async def get_openbis_pat(
+    host: str,
+    session_id: str,
+    personal_access_token_session_name: str = "renku",
+    minimum_validity_in_days: int = 2,
+    timeout: int = 12,
+) -> tuple[str, datetime]:
+    """Requests an openBIS PAT with an openBIS session ID."""
+    url = _get_url(host)
+
+    async with httpx.AsyncClient(verify=get_ssl_context()) as client:
+        get_server_information = {"method": "getServerInformation", "params": [session_id], "id": "2", "jsonrpc": "2.0"}
+        response = await client.post(url, json=get_server_information, timeout=timeout)
+        if response.status_code == 200:
```

Comment:
suggestion: I'd flip the if-statement to `!= 200` with error raising to reduce nesting of this code block and keep the raised error closer to the thing that causes it, same for the if statements below, with more tailored error messages to help tell apart what went wrong if something goes wrong. it's more code to write but if something ever goes wrong it's much easier to debug.

Reply:
I prefer "whitelist" ifs because they make the code more robust. And robust code does not need to be debugged 😉  Anyway, I can apply your suggestions if you want.

Reply:
it's up to you (that's why marked as suggestion). i prefer less nesting but it's a matter of choice :)